### PR TITLE
Upgrade to scala-asm 6.2

### PR DESCRIPTION
**Pull request by retronym**
See: https://github.com/scala/scala-asm/issues/5

Upstream changes in ASM:

  https://github.com/scala/scala-asm/compare/ASM_6_0...ASM_6_2
  http://asm.ow2.io/versions.html

The motivations, other than just keeping current, are:

  - support for Java 9/10/11 updates to the classfile format.
  - reducing needless String => Array[Char] conversions thanks
    to internal changes in ASM.

This PR will fail to build until we publish artifact
from scala/scala-asm.

Fixes scala/bug#10717

### CODE REVIEW

**Comment by retronym** [comment]
Under the `scalap` corpus, this gives us a 3% reduction in allocations. These might be coming from https://gitlab.ow2.org/asm/asm/commit/5db12737cbb662c56bc5ed9cc9902530b3ec1154. I'll compare profiles to learn more.

I'm still measuring effect on compile time in the benchmark.

**Comment by lrytz** [comment]
LGTM, let's release `scala-asm 6.2.0-scala-1`

**Comment by mkurz** [comment]
Maybe it's time to try to get all the patches into the official [asm repo](https://gitlab.ow2.org/asm/asm)?

**Comment by adriaanm** [comment]
> Maybe it's time to try to get all the patches into the official asm repo?

We'd love to. Will keep trying (recent attempt: https://gitlab.ow2.org/asm/asm/issues/317828#note_35917). (EDIT: expanding a bit: they don't seem to have a clear story on how we can contribute our patches, but the bug we reported above was fixed. If someone here has a connection to the ASM maintainers and could introduce us, that'd be great.)

**Review by lrytz — APPROVED** [review]
Squashed the two commits and included a fix for the failing test

**Comment by lrytz** [comment]
Hmm, travis aborted

```
[warn] /home/travis/build/scala/scala/src/library/scala/collection/View.scala:147: @SerialVersionUID has no effect on non-serializable classes
[warn]   class Partition[A](val underlying: SomeIterableOps[A], val p: A => Boolean) {
[warn]         ^
[warn] /home/travis/build/scala/scala/src/library/scala/collection/View.scala:291: @SerialVersionUID has no effect on non-serializable classes
[warn]   class Unzip[A, A1, A2](underlying: SomeIterableOps[A])(implicit asPair: A => (A1, A2)) {
[warn]         ^
[warn] /home/travis/build/scala/scala/src/library/scala/collection/View.scala:297: @SerialVersionUID has no effect on non-serializable classes
[warn]   class Unzip3[A, A1, A2, A3](underlying: SomeIterableOps[A])(implicit asTriple: A => (A1, A2, A3)) {
[warn]         ^
[warn] there were 36 deprecation warnings (since 2.10.0)
[warn] there was one deprecation warning (since 2.11.0)
[warn] there were two deprecation warnings (since 2.12.0)
[warn] there were 14 deprecation warnings (since 2.13.0)
[warn] there were 53 deprecation warnings in total; re-run with -deprecation for details
[warn] there were 11 feature warnings; re-run with -feature for details
[warn] there were 411 inliner warnings; re-run enabling -opt-warnings for details, or try -help
[warn] 10 warnings found
[warn] there were two deprecation warnings
[warn] there were two deprecation warnings (since 2.10.0)
[warn] there were four deprecation warnings (since 2.10.1)
[warn] there were 15 deprecation warnings (since 2.11.0)
[warn] there were two deprecation warnings (since 2.12.3)
[warn] there were 22 deprecation warnings (since 2.13.0)
[warn] there were 47 deprecation warnings in total; re-run with -deprecation for details
[warn] there were four unchecked warnings; re-run with -unchecked for details
[warn] there were four inliner warnings; re-run enabling -opt-warnings for details, or try -help
[warn] 9 warnings found
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
```

and jenkins seems to be stuck in the same place...

**Comment by lrytz** [comment]
Jenkins continued now; but it does look little suspicious.

**Comment by retronym** [comment]
I've seen that output on Travis recently, so it might be a latent problem on this branch (OSGi plugin or something?) I'll take a look locally on Monday.

**Comment by lrytz** [comment]
So you'd like to wait before merging?

**Comment by retronym** [comment]
Yes, let’s hold off

**Comment by retronym** [comment]
Rebased. I tried locally to see if the (bootstrapped) build got stuck in a similar spot to those logs but didn't see it. I'd say we should merge this and revert it we see the problems in CI.

**Comment by retronym** [comment]
Travis [failed again on 2d3d96c](https://travis-ci.org/scala/scala/jobs/394352897) with the same timeout. 😖

**Comment by retronym** [comment]
I've captured [stacks of the slowness in Jenkins](https://gist.github.com/e71869a5499b517dc681a169c477fdf3). Looks like it shows up under `-opt` in `ProdConsAnalyzer`.

**Comment by retronym** [comment]
I've added a benchmark that shows the performance regression, which starts right away in 54dbb0f.

Here are the profiles [before and after](https://drive.google.com/drive/folders/1V1nenx2Hemxidk9d3pgjON_KR8NHXRHq?usp=sharing)

Looks like we're falling into this code much more than before: https://github.com/scala/scala-asm/blob/2e8a7bb0634ed905f7df86ba2ecef08c90855636/src/main/java/scala/tools/asm/tree/analysis/SourceInterpreter.java#L205-L208

**Comment by retronym** [comment]
This [diff in ASM](https://github.com/retronym/asm/commit/70960d92caff136011995b3acd2e04994aa4ad50#diff-704544ff4289963c893e341c0acbf999R275) triggers the slowness.

It now creates a new `Frame` for each instructon ➡️ enclosing-handler edge. We capture this in our pseudo instruction `ExceptionProducer`: https://github.com/scala/scala/blob/54dbb0f80ee1f5b13f8012acb5008b7eba4f3692/src/compiler/scala/tools/nsc/backend/jvm/analysis/ProdConsAnalyzerImpl.scala#L458

These no longer get collapsed when entering them into the merging `SourceValue#insn` sets together in `SourceInterpreter#merge`.

The fix looks straight forward enough: just capture the `handlerFrame.stackTop` in our pseudo instruction, which is all we need later on.

**Comment by retronym** [comment]
@lrytz Since you last looked:

  - I've included "Customise label handling without needing to modify ASM directly"
  - The performance regression in prod/cons is identified and remedied
  - I've added a commit to show that the "call copyOperation consistently" change that _didn't_ get merged to upstream ASM isn't needed anyway, at least accordng to our test suite and my understanding of usage of ASM.

After this, we're ready to move to vanilla ASM 6.3. We still need to decided on whether to depend on it as a regular dependency or keep the status quo of repackaging (either with the scripts in scala-asm or perhaps just with JarJar).

File: src/compiler/scala/tools/nsc/backend/jvm/ClassNode1.java
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/ClassNode1.java]: Le'ts add a comment why this custom class exists

File: src/compiler/scala/tools/nsc/backend/jvm/MethodNode1.java
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/MethodNode1.java]: Also here, a comment would be good.

File: src/compiler/scala/tools/nsc/backend/jvm/MethodNode1.java
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/MethodNode1.java:35]: And here, where the actual change is.

File: src/compiler/scala/tools/nsc/backend/jvm/LabelNode1.java
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/LabelNode1.java]: Add a comment what we customize

File: src/compiler/scala/tools/nsc/backend/jvm/analysis/AliasingFrame.scala
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/analysis/AliasingFrame.scala]: AliasingFrame is used in copy propagation and nullness analysis, but not in ProdConsAnalyzer. The following diff does that: https://github.com/retronym/scala/compare/bump/asm-6.2...lrytz:pr6733?expand=1

With that, some tests `ProdConsAnalyzerTest` actually start failing.

There might be a way to implement prod-cons without those additional calls to `copyOperation`, I haven't looked in detail.

**Review by lrytz — COMMENTED** [review]
Thanks a lot for finding and fixing the performance regression!

File: src/compiler/scala/tools/nsc/backend/jvm/analysis/AliasingFrame.scala
Comment (retronym) [reply src/compiler/scala/tools/nsc/backend/jvm/analysis/AliasingFrame.scala]: I would expect `ProdConsAnalyzer` to fail with the `VanillaAsmFrame` until ASM 6.3 which will start to call `newParameterValue`/`newEmptyNonParameterLocalValue`/`newExceptionValue`, which `InitialProducerSourceInterpreter` implements.

We only override `copyOperation` in `NullnessInterpreter`.

**Comment by retronym** [comment]
I've removed "38826f6320 Show that we don't actually need 'call copyOperation consistently" as it is just code that would be redundant with ASM 6.3. I've also added comments to the `asm.tree._Node` customizations.

File: src/compiler/scala/tools/nsc/backend/jvm/analysis/AliasingFrame.scala
Comment (lrytz) [reply src/compiler/scala/tools/nsc/backend/jvm/analysis/AliasingFrame.scala]: Not sure I follow.. The `newXValue` methods are called in the Analyzer, not in `Frame` / `VanillaAsmFrame`. The `copyOperation` implementation for ProdCons is inherited from `SourceInterpreter`. Anyway, we'll see when we update to 6.3.

**Review by lrytz — APPROVED** [review]

File: src/compiler/scala/tools/nsc/backend/jvm/analysis/AliasingFrame.scala
Comment (retronym) [reply src/compiler/scala/tools/nsc/backend/jvm/analysis/AliasingFrame.scala]: Right, I didn't consider the inherited implementation of `copyOperation`, i just looked for the one override in our codebase. Still, it should be easy enough to adapt when we need to.

**Pull request merged**
